File: data_structures/tree/bst.dart
// binary_search_tree.dart

class Node {
  dynamic data;
  Node? left;
  Node? right;

  Node(this.data) : left = null, right = null;
}

class BinarySearchTree {
  Node? root;

  BinarySearchTree() : root = null;

  void insert(dynamic data) {
    root = _insertRecursive(root, data);
  }

  Node? _insertRecursive(Node? root, dynamic data) {
    if (root == null) {
      return Node(data);
    }

    if (data < root.data) {
      root.left = _insertRecursive(root.left, data);
    } else if (data > root.data) {
      root.right = _insertRecursive(root.right, data);
    }

    return root;
  }

  bool search(dynamic data) {
    return _searchRecursive(root, data);
  }

  bool _searchRecursive(Node? root, dynamic data) {
    if (root == null) {
      return false;
    }

    if (data == root.data) {
      return true;
    } else if (data < root.data) {
      return _searchRecursive(root.left, data);
    } else {
      return _searchRecursive(root.right, data);
    }
  }

  void inOrderTraversal() {
    _inOrderTraversalRecursive(root);
  }

  void _inOrderTraversalRecursive(Node? root) {
    if (root != null) {
      _inOrderTraversalRecursive(root.left);
      print(root.data);
      _inOrderTraversalRecursive(root.right);
    }
  }
}

void main() {
  BinarySearchTree bst = BinarySearchTree();

  bst.insert(5);
  bst.insert(3);
  bst.insert(7);
  bst.insert(2);
  bst.insert(4);
  bst.insert(6);

  print("In-Order Traversal:");
  bst.inOrderTraversal();

  print("\nSearch for 4: ${bst.search(4)}");
  print("Search for 8: ${bst.search(8)}");
}

File: go/data_structures/tree/bst.go
// bst.go

package main

import "fmt"

type BSTNode struct {
    data  int
    left  *BSTNode
    right *BSTNode
}

type BST struct {
    root *BSTNode
}

func NewBST() *BST {
    return &BST{}
}

func (bst *BST) Insert(data int) {
    bst.root = insertNode(bst.root, data)
}

func insertNode(root *BSTNode, data int) *BSTNode {
    if root == nil {
        return &BSTNode{data: data}
    }

    if data < root.data {
        root.left = insertNode(root.left, data)
    } else if data > root.data {
        root.right = insertNode(root.right, data)
    }

    return root
}

func (bst *BST) Display() {
    fmt.Print("DFS Inorder Traversal: ")
    inorderTraversal(bst.root)
    fmt.Println()

    fmt.Print("DFS Preorder Traversal: ")
    preorderTraversal(bst.root)
    fmt.Println()

    fmt.Print("DFS Postorder Traversal: ")
    postorderTraversal(bst.root)
    fmt.Println()

    fmt.Print("BFS Level Order Traversal: ")
    levelOrderTraversal(bst.root)
    fmt.Println()
}

func inorderTraversal(root *BSTNode) {
    if root != nil {
        inorderTraversal(root.left)
        fmt.Printf("%d ", root.data)
        inorderTraversal(root.right)
    }
}

func preorderTraversal(root *BSTNode) {
    if root != nil {
        fmt.Printf("%d ", root.data)
        preorderTraversal(root.left)
        preorderTraversal(root.right)
    }
}

func postorderTraversal(root *BSTNode) {
    if root != nil {
        postorderTraversal(root.left)
        postorderTraversal(root.right)
        fmt.Printf("%d ", root.data)
    }
}

func levelOrderTraversal(root *BSTNode) {
    if root == nil {
        return
    }

    queue := []*BSTNode{root}
    for len(queue) > 0 {
        currentNode := queue[0]
        queue = queue[1:]

        fmt.Printf("%d ", currentNode.data)

        if currentNode.left != nil {
            queue = append(queue, currentNode.left)
        }

        if currentNode.right != nil {
            queue = append(queue, currentNode.right)
        }
    }
}

func main() {
    bst := NewBST()
    bst.Insert(4)
    bst.Insert(2)
    bst.Insert(6)
    bst.Insert(1)
    bst.Insert(3)
    bst.Insert(5)
    bst.Insert(7)

    fmt.Println("BST:")
    bst.Display()
}
